### ISSUE TITLE
Support alternative remote playpens for code execution

### ISSUE THREAD

**Issue by budziq** [azerupi/mdBook#350]
Like:
- https://play.integer32.com/ for rust
- https://rust.godbolt.org/ - compiler explorer is not a playpen pe se but it's immensely useful. (Also for other languages)

Also take into consideration support for other languages:
- https://play.golang.org/
- https://code.sololearn.com/#cpp

See https://fiddles.io/ for extended list

**Comment by agauniyal** [comment]
+1 for godbolt. I would like to present [wandbox](https://github.com/melpon/wandbox/blob/master/kennel2/API.rst) as common denominator here as well.

**Comment by CjS77** [comment]
There's an easy workaround here:
mdbook init --theme

and edit book.js to point to your custom playpen/playground (Rust only, mind you)

**Comment by vanceism7** [comment]
I'm really interested in getting this feature going - Mdbook is great but it'd be nice if we could hook into other playground code editors (like purescript, haskell, etc). I don't mind trying to implement this myself, does anyone have any suggestions as a good starting point?

**Comment by julianfortune** [comment]
@vanceism7 Haskell folks have created a public API for executing code:
- https://play.haskell.org 
- https://github.com/haskell/play-haskell#api

**Comment by Kleidukos** [comment]
Indeed, and we're also looking into integrating the Haskell playground into mdBook. :)

**Comment by kwest-ng** [comment]
I'm quite interested in implementing this, for my own book. I could get started on this now, and I have a decent idea about how to do it, as the codebase is not terribly complex.  But I would prefer to have a reasonable idea that this is a desired change before committing too much time to it.

The rough idea I had was two-fold:

- Provide a clear JS API for interacting with other playgrounds, populated by the config from `book.toml`.
- Create configuration in `book.toml`, as well as defaults for rust that map to the current playground settings. Appropriate config params will be passed to the JS API.

The simplest generalized solution I can see for the JS API is to split the [`run_rust_code`](https://github.com/rust-lang/mdBook/blob/30e0e1d1021d406899c783447994de0a916643c3/crates/mdbook-html/front-end/js/book.js#L105) function into two parts:

- Identifying the language, reading the code, and editing the DOM with the resulting text (whether or not it was successful). This part will likely be the same for all languages, and can be used as a wrapper for calling the second part of the JS API. This code will be part of `mdBook`.
- Provided with code text and config parameters, perform whatever task is needed to produce output. This may include arbitrary execution, including HTTP requests, importing/calling WASM, etc. Output should be returned from the function as a string.

Proposed changes would look something like the following:
```toml
# book.toml

# Name must match the language name in the markdown code block,
# that's how we'll detect which playground to run.
[playground.haskell]
source = "path/to/run_haskell.js" # file name is arbitrary

# This block should get passed directly to the custom JS function
[playground.haskell.extra]
url = "https://play.haskell.org/submit"
```

```js
// path/to/run_haskell.js

// language name in function must mach code block language name
async function run_haskell_playground(code, config) {
  // example code, this is arbitrary and provided by the user.
  let url = config.url;
  let request = prepare_request(code, url);
  let result = await fetch(request);
  return extract_response_from(result);
}
```

`book.js` will need to become a template, rather than a standalone js asset.
```js
// book.js.hbs

// import run_haskell_playground from source file
// alternatively, this could be imported as a <script> in index.hbs

// This is the common entry point for all playground code
// but would take the place of the current `run_rust_code`
function run_playground_code(
  // verify we have a function to call
  // Ingest config from `extra` block (ideally this would be done at build time)
  // Ingest code text
  // set up result DOM element with a "Running..." message;
  let result = run_haskell_playground(code, config);
  // verify result is sensible (i.e. is a string)
  // insert result string into the DOM element set up earlier
)
```

### Open questions

- Should `mdBook` simply own these files, or require the book author to define them?
  - If the author should define them, how does the rust playground get defined? It should still be built-in and enabled by default, but can that be disabled/overridden?
- Are there any additional items that must be passed to the custom function, but can't be handled by `book.toml`?
- sync vs `async`?

**Comment by kwest-ng** [comment]
Under that idea, the rust-specific code would look something like:
```js
async function run_rust_playground(code, config) {
    // how to handle edition selecting? maybe a `classes` param as well?
    const classes = code_block.querySelector('code').classList;
    let edition = '2015';
    classes.forEach(className => {
        if (className.startsWith('edition')) {
            edition = className.slice(7);
        }
    });
    const params = {
        version: 'stable',
        optimize: '0',
        code: text,
        edition: edition,
    };

    if (text.indexOf('#![feature') !== -1) {
        params.version = 'nightly';
    }

    // Are users required to use this function to make HTTP calls?
    // Is it a footgun to not provide this?
    let response = await fetch_with_timeout('https://play.rust-lang.org/evaluate.json', {
        headers: {
            'Content-Type': 'application/json',
        },
        method: 'POST',
        mode: 'cors',
        body: JSON.stringify(params),
    };
    // Errors do not need to be handled in this function, the wrapper will handle them
    // Functions may opt in to manual error handling to improve messages
    return response.json().result;
}
```
